src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1264 @@
+    idc_enbl = !!(vaplic_get_idelivery(vcpu, vcpu->id));
+    idc_force = !!(vaplic_get_iforce(vcpu, vcpu->id));
+      
+    if ((intp_id != APLIC_MAX_INTERRUPTS) && 
+        (intp_prio < idc_threshold || idc_threshold == 0) && 
+            idc_enbl && domain_enbl) {
+            update_topi = (intp_id << 16) | intp_prio;
+            ret = true;
+    } else if (idc_force && idc_enbl && domain_enbl) {
+        ret = true;
+    }
+    vaplic->topi_claimi[vcpu->id] = update_topi;
+    return ret;
+}
+
+enum {UPDATE_HART_LINE};
+static void vaplic_ipi_handler(uint32_t event, uint64_t data);
+CPU_MSG_HANDLER(vaplic_ipi_handler, VPLIC_IPI_ID);
+
+/**
+ * @brief Updates the interrupt line for a single hart
+ * 
+ * @param vcpu virtual cpu
+ * @param vhart_index hart id to update
+ */
+static void vaplic_update_hart_line(struct vcpu* vcpu, vcpuid_t vhart_index){
+    cpuid_t pcpu_id = vaplic_vcpuid_to_pcpuid(vcpu, vhart_index);
+
+    /** 
+     *  If the current cpu is the targeting cpu, signal the intp 
+     *  to the hart
+     *  Else, send a mensage to the targeting cpu 
+     */
+    if(pcpu_id == cpu()->id) {
+        if(vaplic_update_topi(vcpu)){
+            CSRS(CSR_HVIP, HIP_VSEIP);
+        } else  {
+            CSRC(CSR_HVIP, HIP_VSEIP);
+        }
+    } else {
+        struct cpu_msg msg = {VPLIC_IPI_ID, UPDATE_HART_LINE, vhart_index};
+        cpu_send_msg(pcpu_id, &msg);       
+    }
+}
+
+/**
+ * @brief Triggers the hart/harts interrupt line update.
+ * 
+ * @param vcpu virtual cpu
+ * @param vhart_index virtual hart to update the interrupt line. 
+ *        If UPDATE_ALL_HARTS were passed, thsi function will trigger
+ *        the interrupt line update to all virtual harts running in this vm.  
+ */
+static void vaplic_update_hart(struct vcpu* vcpu, int16_t vhart_index) 
+{
+    struct vaplic *vaplic = &vcpu->vm->arch.vaplic;
+
+    if (vhart_index == UPDATE_ALL_HARTS){
+        for(size_t i = 0; i < vaplic->idc_num; i++){
+            vaplic_update_hart_line(vcpu, (vcpuid_t)i);
+        }
+    } else if (vhart_index < vaplic->idc_num){
+        vaplic_update_hart_line(vcpu, (vcpuid_t)vhart_index);
+    }
+}
+
+/**
+ * @brief Processes an incoming event.
+ * 
+ * @param event the event id
+ * @param data
+ */
+static void vaplic_ipi_handler(uint32_t event, uint64_t data) 
+{
+    switch(event) {
+        case UPDATE_HART_LINE:
+            vaplic_update_hart(cpu()->vcpu, (int16_t)data);
+            break;
+    }
+}
+
+/**
+ * @brief Write to domaincfg register a new value.
+ * 
+ * @param vcpu 
+ * @param new_val The new value to write to domaincfg
+ */
+static void vaplic_set_domaincfg(struct vcpu *vcpu, uint32_t new_val){
+    struct vaplic * vaplic = &vcpu->vm->arch.vaplic;
+    spin_lock(&vaplic->lock);
+    /** Update only the virtual domaincfg */
+    /** Only Interrupt Enable and Delivery mode are configurable */
+    new_val &= (APLIC_DOMAINCFG_IE | APLIC_DOMAINCFG_DM);
```

Comment:
We should probably remove the DM bit configurability and make it hard-coded based on the IRQC configuration, right?

Reply:
Yup. I completely agree. The common case will be DM = 1 when imsics are present, 0 otherwise. There is no reason to allow of a mix of the two in our implementation, it will just add complexity unecessarily.

Reply:
Done!